File: app/src/main/java/me/ykrank/s1next/view/fragment/EditPostFragment.java
package me.ykrank.s1next.view.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.disposables.Disposable;
import me.ykrank.s1next.App;
import me.ykrank.s1next.R;
import me.ykrank.s1next.data.api.S1Service;
import me.ykrank.s1next.data.api.model.Post;
import me.ykrank.s1next.data.api.model.PostEditor;
import me.ykrank.s1next.data.api.model.Thread;
import me.ykrank.s1next.data.api.model.ThreadType;
import me.ykrank.s1next.databinding.FragmentEditPostBinding;
import me.ykrank.s1next.util.ErrorUtil;
import me.ykrank.s1next.util.L;
import me.ykrank.s1next.util.RxJavaUtil;
import me.ykrank.s1next.view.adapter.SimpleSpinnerAdapter;
import me.ykrank.s1next.view.dialog.EditPostRequestDialogFragment;

/**
 * A Fragment shows {@link EditText} to let the user enter reply.
 */
public final class EditPostFragment extends BasePostFragment {

    public static final String TAG = EditPostFragment.class.getName();

    private static final String ARG_THREAD = "thread";
    private static final String ARG_POST = "post";

    @Inject
    S1Service mS1Service;

    private Thread mThread;
    private Post mPost;
    private boolean isHost;

    private Disposable mDisposable;

    private FragmentEditPostBinding binding;

    public static EditPostFragment newInstance(@NonNull Thread thread, @NonNull Post post) {
        EditPostFragment fragment = new EditPostFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(ARG_THREAD, thread);
        bundle.putParcelable(ARG_POST, post);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentEditPostBinding.inflate(inflater, container, false);
        initCreateView(binding.layoutPost);

        mThread = getArguments().getParcelable(ARG_THREAD);
        mPost = getArguments().getParcelable(ARG_POST);

        isHost = mPost != null && mPost.isFirst();
        binding.setHost(isHost);
        L.leaveMsg(String.format("EditPostFragment##post:%s", mPost));

        return binding.getRoot();
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        App.getAppComponent().inject(this);
        init();
    }

    @Override
    public void onDestroy() {
        RxJavaUtil.disposeIfNotNull(mDisposable);
        super.onDestroy();
    }

    @Override
    protected boolean OnMenuSendClick() {
        String typeId = null;
        if (binding.spinner.getVisibility() == View.VISIBLE) {
            ThreadType selectType = (ThreadType) binding.spinner.getSelectedItem();
            if (selectType == null) {
                showShortSnackbar(R.string.error_not_init);
                return true;
            }
            typeId = selectType.getTypeId();
            //未选择类别
            if (typeId == null || "0".equals(typeId.trim())) {
                showShortSnackbar(R.string.error_no_type_id);
                return true;
            }
        }

        String title = binding.title.getText().toString();
        String message = mReplyView.getText().toString();
        if (!isTitleValid(title) || !isMessageValid(message)) {
            showShortSnackbar(R.string.error_no_title_or_message);
            return true;
        }

        EditPostRequestDialogFragment.newInstance(mThread, mPost, typeId, title, message)
                .show(getFragmentManager(), EditPostRequestDialogFragment.TAG);

        return true;
    }

    @Nullable
    @Override
    public String getCacheKey() {
        return null;
    }

    private boolean isTitleValid(String string) {
        if (!isHost) {
            return true;
        }
        if (TextUtils.isEmpty(string) || string.trim().length() == 0) {
            return false;
        }
        return true;
    }

    private boolean isMessageValid(String string) {
        if (TextUtils.isEmpty(string) || string.trim().length() == 0) {
            return false;
        }
        return true;
    }

    private void init() {
        mDisposable = mS1Service.getEditPostInfo(mThread.getFid(), mThread.getId(), mPost.getId())
                .map(PostEditor::fromHtml)
                .compose(RxJavaUtil.iOTransformer())
                .subscribe(postEditor -> {
                    if (isHost) {
                        setSpinner(postEditor.getThreadTypes());
                        binding.spinner.setSelection(postEditor.getTypeIndex());
                        binding.title.setText(postEditor.getSubject());
                    }
                    binding.layoutPost.reply.setText(postEditor.getMessage());
                }, e -> {
                    L.report(e);
                    showRetrySnackbar(ErrorUtil.parse(getContext(), e), v -> init());
                });

    }

    private void setSpinner(@Nullable List<ThreadType> types) {
        if (types == null || types.isEmpty()) {
            binding.spinner.setVisibility(View.GONE);
            return;
        } else {
            binding.spinner.setVisibility(View.VISIBLE);
        }
        SimpleSpinnerAdapter<ThreadType> spinnerAdapter = new SimpleSpinnerAdapter<>(getContext(), types, ThreadType::getTypeName);
        binding.spinner.setAdapter(spinnerAdapter);
    }

}

File: app/src/main/java/me/ykrank/s1next/view/fragment/EditPostFragment.kt
package me.ykrank.s1next.view.fragment

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import com.github.ykrank.androidautodispose.AndroidRxDispose
import com.github.ykrank.androidlifecycle.event.FragmentEvent
import com.github.ykrank.androidtools.util.L
import com.github.ykrank.androidtools.util.RxJavaUtil
import me.ykrank.s1next.App
import me.ykrank.s1next.R
import me.ykrank.s1next.data.api.S1Service
import me.ykrank.s1next.data.api.model.Post
import me.ykrank.s1next.data.api.model.PostEditor
import me.ykrank.s1next.data.api.model.Thread
import me.ykrank.s1next.data.api.model.ThreadType
import me.ykrank.s1next.databinding.FragmentEditPostBinding
import me.ykrank.s1next.view.adapter.SimpleSpinnerAdapter
import me.ykrank.s1next.view.dialog.requestdialog.EditPostRequestDialogFragment
import me.ykrank.s1next.view.event.RequestDialogSuccessEvent
import me.ykrank.s1next.view.page.post.postedit.BasePostEditFragment
import javax.inject.Inject

/**
 * A Fragment shows [EditText] to let the user enter reply.
 */
class EditPostFragment : BasePostEditFragment() {

    @Inject
    internal lateinit var mS1Service: S1Service

    private lateinit var mThread: Thread
    private lateinit var mPost: Post
    private var isHost: Boolean = false

    private lateinit var binding: FragmentEditPostBinding

    override fun onAttach(context: Context) {
        super.onAttach(context)
        App.appComponent.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentEditPostBinding.inflate(inflater, container, false)
        initCreateView(binding.layoutPost)

        mThread = requireArguments().getParcelable(ARG_THREAD)!!
        mPost = requireArguments().getParcelable(ARG_POST)!!

        isHost = mPost.isFirst
        binding.host = isHost
        leavePageMsg(String.format("EditPostFragment##post:%s", mPost))

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        init()
    }

    override fun onMenuSendClick(): Boolean {
        var typeId: String? = null
        if (binding.spinnerType.visibility == View.VISIBLE) {
            val selectType = binding.spinnerType.selectedItem as ThreadType?
            if (selectType == null) {
                showSnackbar(R.string.error_not_init)
                return true
            }
            typeId = selectType.typeId
            //未选择类别
            if (typeId == null || "0" == typeId.trim()) {
                showSnackbar(R.string.error_no_type_id)
                return true
            }
        }
        var readPerm: String? = null
        if (binding.spinnerPerm.visibility == View.VISIBLE) {
            readPerm = binding.spinnerPerm.selectedItem as String?
        }

        val title = binding.title.text?.toString()
        val message = mReplyView.text?.toString()
        if (!isTitleValid(title) || !isMessageValid(message)) {
            showSnackbar(R.string.error_no_title_or_message)
            return true
        }

        EditPostRequestDialogFragment.newInstance(mThread, mPost, typeId, readPerm, title!!, message!!)
                .show(parentFragmentManager, EditPostRequestDialogFragment.TAG)

        return true
    }

    override val cacheKey: String?
        get() = null

    private fun isTitleValid(string: String?): Boolean {
        if (!isHost) {
            return true
        }
        if (string.isNullOrBlank()) {
            return false
        }
        return true
    }

    private fun isMessageValid(string: String?): Boolean {
        if (string.isNullOrBlank()) {
            return false
        }
        return true
    }

    private fun init() {
        mS1Service.getEditPostInfo(mThread.fid!!.toInt(), mThread.id!!.toInt(), mPost.id)
                .map<PostEditor> { PostEditor.fromHtml(it) }
                .compose(RxJavaUtil.iOSingleTransformer())
                .to(AndroidRxDispose.withSingle(this, FragmentEvent.DESTROY))
                .subscribe({ postEditor ->
                    if (isHost) {
                        setSpinnerType(postEditor.threadTypes)
                        binding.spinnerType.setSelection(postEditor.typeIndex)

                        setSpinnerPerm(postEditor.readPermTypes)
                        binding.spinnerPerm.setSelection(postEditor.readPermIndex)

                        binding.title.setText(postEditor.subject)
                    }
                    binding.layoutPost.reply.setText(postEditor.message)
                }, { e ->
                    L.report(e)
                    showRetrySnackbar(e, View.OnClickListener { init() })
                })

    }

    private fun setSpinnerType(types: List<ThreadType>?) {
        if (types.isNullOrEmpty()) {
            binding.spinnerType.visibility = View.GONE
            return
        } else {
            binding.spinnerType.visibility = View.VISIBLE
        }
        val spinnerAdapter = SimpleSpinnerAdapter(requireContext(), types) { it?.typeName.toString() }
        binding.spinnerType.adapter = spinnerAdapter
    }

    private fun setSpinnerPerm(types: List<String>?) {
        if (types.isNullOrEmpty()) {
            binding.spinnerPerm.visibility = View.GONE
            return
        } else {
            binding.spinnerPerm.visibility = View.VISIBLE
        }
        val spinnerAdapter = SimpleSpinnerAdapter(requireContext(), types) { it.toString() }
        binding.spinnerPerm.adapter = spinnerAdapter
    }

    override fun isRequestDialogAccept(event: RequestDialogSuccessEvent): Boolean {
        return event.dialogFragment is EditPostRequestDialogFragment
    }

    override fun onRequestDialogSuccess() {
        activity?.setResult(Activity.RESULT_OK)
    }

    companion object {

        val TAG: String = EditPostFragment::class.java.simpleName

        private const val ARG_THREAD = "thread"
        private const val ARG_POST = "post"

        fun newInstance(thread: Thread, post: Post): EditPostFragment {
            val fragment = EditPostFragment()
            val bundle = Bundle()
            bundle.putParcelable(ARG_THREAD, thread)
            bundle.putParcelable(ARG_POST, post)
            fragment.arguments = bundle
            return fragment
        }
    }
}
